saleor/discount/utils.py

```diff
@@ -887,39 +1120,63 @@ def _handle_gift_reward(
         if fields_to_update:
             line_discount.save(update_fields=fields_to_update)
 
-    checkout_info.discounts = []
-    checkout.discount_amount = Decimal("0")
-    if save:
-        checkout.save(update_fields=["discount_amount"])
+    if checkout_info:
+        checkout_info.discounts = []
+        checkout_info.checkout.discount_amount = Decimal("0")
+        if save:
+            checkout_info.checkout.save(update_fields=["discount_amount"])
 
     if line_created:
         variant = gift_listing.variant
-        gift_line_info = CheckoutLineInfo(
-            line=line,
-            variant=variant,
-            channel_listing=gift_listing,
-            product=variant.product,
-            product_type=variant.product.product_type,
-            collections=[],
-            discounts=[line_discount],
-            rules_info=[rule_info],
-            channel=checkout_info.channel,
-        )
+        init_values = {
+            "line": line,
+            "variant": variant,
+            "channel_listing": gift_listing,
+            "discounts": [line_discount],
+            "rules_info": [rule_info],
+            "channel": order_or_checkout.channel,
+        }
+        if checkout_info:
+            init_values |= {
+                "product": variant.product,
+                "product_type": variant.product.product_type,
+                "collections": [],
+            }
+        gift_line_info = models.line_info(**init_values)
         lines_info.append(gift_line_info)  # type: ignore[attr-defined]
     else:
-        line_info = find_checkout_line_info(lines_info, line.id)
+        line_info = next(
+            line_info for line_info in lines_info if line_info.line.pk == line.id
+        )
         line_info.line = line
         line_info.discounts = [line_discount]
 
 
-def create_gift_line(checkout: "Checkout", variant_id: int):
-    defaults = {
-        "variant_id": variant_id,
-        "quantity": 1,
-        "currency": checkout.currency,
-    }
-    line, created = CheckoutLine.objects.get_or_create(
-        checkout=checkout, is_gift=True, defaults=defaults
+def _get_defaults_for_line(order_or_checkout: Union[Checkout, Order], variant_id: int):
```

Comment:
The function definition should be below the call, it's easier to read.
Also, I will suggest changing the name to `_get_defaults_for_gift_line`